lib/editor/components/map/LocationsLayer.js

```diff
@@ -0,0 +1,221 @@
+// @flow
+
+import * as L from 'leaflet'
+import clone from 'lodash/cloneDeep'
+import React, { PureComponent } from 'react'
+import { EditControl } from 'react-leaflet-draw'
+import {FeatureGroup, Polygon, Polyline} from 'react-leaflet'
+
+import * as activeActions from '../../actions/active'
+import type { GtfsLocation, LocationShape } from '../../../types'
+import { convertLatLngToArray, groupLocationShapePoints } from '../../util/location'
+
+type GroupedLocationShapePts = {
+  [string]: Array<[number, number]>
+}
+
+type Props = {
+  activeEntity: GtfsLocation,
+  updatingLocationShape: typeof activeActions.updatingLocationShape
+}
+
+type State = {
+  initialShapeIds: Array<string>
+}
+
+/**
+ * Draw all locations on a map and render a popup on click
+ */
+export default class LocationsLayer extends PureComponent<Props, State> {
+  state = {
+    initialShapeIds: []
+  }
+
+  constructor (props: Props) {
+    super(props)
+    const locationShapes = this.props.activeEntity.location_shapes
+    // Extract all the unique geometry IDs out of the location shape points.
+    this.state.initialShapeIds = [...new Set(locationShapes.map(shapePt => shapePt.geometry_id))]
+  }
+
+  checkSameShape = (shape: Array<[number, number]>, layerCoords: Array<[number, number]>) => {
+    // We receive layerCoords in an array of coordinates [[YY.YYY, ZZ.ZZZZ], [..]]
+    // We then package geometryCoords back into a similar array to compare
+    // TODO: this function is not set up to handle the case where coords represents a single polygon
+    // and shape represents multiple (i.e. MultiPolygons)
+    // const geometryCoords = shape.forEach(shapePt => [shapePt.geometry_pt_lat, shapePt.geometry_pt_lon])
+    return JSON.stringify(shape) === JSON.stringify(layerCoords)
+  }
+
+  _onDrawStart (e: L.LeafletEvent) {
+    const {layerType} = e
+    const {geometry_type: geometryType} = this.props.activeEntity
+
+    if (geometryType && layerType !== geometryType) {
+      window.alert('Mixed location geometry types are not supported by the GTFS flex specification.')
+    }
+  }
+
+  _onCreated (e: L.LeafletEvent) {
+    const {layer, layer: {_leaflet_id: leafletId}, layerType: geometryType} = e
+    const {activeEntity} = this.props
+
+    // TODO: allow multi polygon type
+    if (activeEntity.location_shapes.length > 0) {
+      window.alert('MultiPolygons / Multipolylines are not currently supported.')
+      // TODO: clear the EditControl in a better way
+      location.reload()
+      return
+    }
+
+    // TODO: unify geometryType w/ backend types, including detecting MULTIPOLYGON & MULTILINESTRING
+    // Construct a location type as an array of location shape points.
+    const locationShape: LocationShape = geometryType === 'polygon'
+      ? layer.getLatLngs()[0].map(latlng => {
```

Comment:
Same remark as https://github.com/ibi-group/datatools-ui/pull/754#discussion_r798778538.

Reply:
The link doesn't work for me. I'm guessing it's about moving this to a util function? I can't see a copy of this code